File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/security/token/JwtRefreshToken.java
package com.ntnu.gidd.security.token;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jws;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.List;
import java.util.Optional;


/**
 * Implementation of {@link JwtToken} for refresh tokens.
 */
@Getter
@AllArgsConstructor(access = AccessLevel.PROTECTED)
public class JwtRefreshToken implements JwtToken {

    private String token;
    private Jws<Claims> claims;

    public static Optional<JwtRefreshToken> of(RawJwtAccessToken token, String signingKey) {
        Jws<Claims> claims = token.parseClaims(signingKey);

        List<String> scopes = claims.getBody()
                .get("scopes", List.class);
        boolean hasRefreshTokenScope = scopes != null && scopes.stream()
                .anyMatch(Scopes.REFRESH_TOKEN.scope()::equals);

        if (hasRefreshTokenScope)
            return Optional.of(new JwtRefreshToken(token.getToken(), claims));

        return Optional.empty();
    }

    @Override
    public String getToken() {
        return token;
    }

    public String getSubject() {
        return claims.getBody().getSubject();
    }

    public String getJti() {
        return claims.getBody().getId();
    }
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/security/token/JwtRefreshToken.kt
package ntnu.idatt2105.security.token

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jws

data class JwtRefreshToken(
    private val token: String,
    private val claims: Jws<Claims>
) : JwtToken {

    companion object {
        fun of(token: RawJwtAccessToken, signingKey: String?): JwtRefreshToken? {
            val claims: Jws<Claims>? = token.parseClaims(signingKey)!!
            val scopes: MutableList<*>? = claims?.getBody()?.get("scopes", MutableList::class.java)
            val hasRefreshTokenScope = scopes != null && scopes.stream()
                    .anyMatch(Scopes.REFRESH_TOKEN.scope()::equals)
            return if (hasRefreshTokenScope) JwtRefreshToken(token.getToken()!!, claims!!) else null
        }
    }

    override fun getToken(): String {
        return token
    }

    fun getSubject(): String {
        return claims.getBody().getSubject()
    }

    fun getJti(): String {
        return claims.getBody().getId()
    }
}
